### ISSUE TITLE
Functions useEmulator not persistent between regions

### ISSUE THREAD

**Issue by epodol** [firebase/firebase-tools#3537]
### [REQUIRED] Environment info

**firebase-tools:** 9.14.0

**Platform:** Windows with WSL 2 Ubuntu

### [REQUIRED] Test case

React and Reactfire are being used in this project. The repo is available at https://github.com/epodol/bsclibrary.

```javascript
        return isDev
          ? functions('us-west2').useEmulator('localhost', 5001)
          : functions('us-west2');
```
```javascript
import { useAuth, useFunctions} from 'reactfire';
// ...
const functions = useFunctions();
// ...
exports.addNewUser = functions
  .region('us-west2')
  .https.onCall(async (data, context) => {
```
### [REQUIRED] Steps to reproduce

1. Set up a firebase project (as normal)
2. Create a callable function in another region (us-west2 is used in this example)
3. Call `functions()` with a custom region as depicted here: https://firebase.google.com/docs/functions/locations#client-side_location_selection_for_callable_functions
4. When tagging a `useEmulator()` call onto the `functions()` with a region, it will result in both the region and the emulator call being ignored.

### [REQUIRED] Expected behavior

The emulator should still work when a custom region is defined.

### [REQUIRED] Actual behavior
```javascript
functions('us-west2').useEmulator('localhost', 5001)
```
![image](https://user-images.githubusercontent.com/34190213/123661557-7e39c880-d7e9-11eb-8006-50c3ec206780.png)
```javascript
functions().useEmulator('localhost', 5001)
```
![image](https://user-images.githubusercontent.com/34190213/123664688-6b74c300-d7ec-11eb-8dba-ee0df2878656.png)
![image](https://user-images.githubusercontent.com/34190213/123677187-ac270900-d7f9-11eb-96c1-3ae32344a812.png)

![image](https://user-images.githubusercontent.com/34190213/123690563-5eb29800-d809-11eb-93ee-2d3068b88751.png)

May be related to #3031, #3364, #3418

**Comment by i14h** [comment]
thanks @epodol for reporting this. @kmcnellis can you please take a look at this?

**Comment by epodol** [comment]
I have looked into this some more and created a basic html/js only example. Here are a few and some conclusions I came to.

## Examples
<details>
<summary>Example 1:</summary>
<br>

```javascript
    firebase.initializeApp(firebaseConfig);

    const functions = firebase.functions();
    functions.useEmulator('localhost', 5001)
    functions.httpsCallable('test')()
```
Example 1 calls: ` http://localhost:5001/$PROJECT/us-central1/test`
</details>
<details>
<summary>Example 2:</summary>
<br>

```javascript
    firebase.initializeApp(firebaseConfig);

    const functions = firebase.app().functions('us-west2');
    functions.useEmulator('localhost', 5001)
    functions.httpsCallable('test')()
```
Example 2 calls `http://localhost:5001/$PROJECT/us-west2/test` (Ideal)
</details>
<details>
<summary>Example 3:</summary>
<br>

```javascript
    firebase.initializeApp(firebaseConfig);

    const functions = firebase.app().functions('us-west2');
    firebase.app().functions().useEmulator('localhost', 5001)
    firebase.app().functions().httpsCallable('test')()
```
Example 3 calls `http://localhost:5001/$PROJECT/us-central1/test` (Back to example 1)
</details>
<details>
<summary>Example 4:</summary>
<br>

```javascript
    firebase.initializeApp(firebaseConfig);

    const functions = firebase.app().functions('us-west2');
    firebase.app().functions('us-west2').useEmulator('localhost', 5001)
    firebase.app().functions('us-west2').httpsCallable('test')()
```
Example 4 calls `http://localhost:5001/$PROJECT/us-west2/test` (Ideal)
</details>
<details>
<summary>Example 5:</summary>
<br>

```javascript
    firebase.initializeApp(firebaseConfig);

    const functions = firebase.app().functions('us-west2');
    functions.useEmulator('localhost', 5001)
    firebase.app().functions().httpsCallable('test')()
```
Example 5 calls `https://us-central1-$PROJECT.cloudfunctions.net/test` (Calling Production)
</details>
<details>
<summary>Example 6:</summary>
<br>

```javascript
    firebase.initializeApp(firebaseConfig);

    const functions = firebase.app().functions('us-west2');
    functions.useEmulator('localhost', 5001)
    firebase.app().functions('us-west2').httpsCallable('test')()
```
Example 6 calls `http://localhost:5001/$PROJECT/us-west2/test` (Ideal)
</details>

## Conclusions

Calling `useEmulator(...)` in Example 5 does not do anything and will always call production, all the other examples change behavior based on the `useEmulator(...)` call. The only difference between Examples 5 and 6 is that the `us-west2` region is defined in Example 6 right before calling the callable function.

I'm not sure if this is actually a bug with Firebase anymore, but I think there could be some improvement with how Firebase handles custom regions. Ideally, the region could be defined as an option/argument in the same `httpsCallable()` call so functions with different regions could be called without having to initialize `functions()` again.

Currently, the best solution for me is to use Example 6 since I am using reactfire. For example:

```javascript
import { useAuth, useFirebaseApp } from 'reactfire';
// ...
  const functions = useFirebaseApp().functions('us-west2');
// ...
          await functions
            .httpsCallable('addNewUser')({
// ...
```

If this is all intended behavior, then you can close this issue/change the title. Thank you!

**Comment by kmcnellis** [comment]
I think @samtstern is the most familiar with this code.

**Comment by kmcnellis** [comment]
But I think basically the issue is in how the useEmulator function interacts with the firebase.app().functions().

The example#5 above, calling production:
```    
    firebase.initializeApp(firebaseConfig);
    const functions = firebase.app().functions('us-west2');
    functions.useEmulator('localhost', 5001)
    firebase.app().functions().httpsCallable('test')()
```

The line `const functions = firebase.app().functions('us-west2');` creates an instance of the functions class, which is then modified by the call to `useEmulator` (defined [here](https://github.com/firebase/firebase-js-sdk/blob/5d31e2192d0ea68a768bc7826ad5aa830c2bc36c/packages/functions/src/api/service.ts#L163) ) to use the emulator origin.  However, it does not change any SDK-wide variables.  

Additionally, there is a layer of caching of the functions instance using the [ComponentContainer](https://github.com/firebase/firebase-js-sdk/blob/919413cd06478365e306476c8eb2ad2aa1f2d3ee/packages/component/src/component_container.ts#L25), that will return that same 'us-west2' instance of functions each time `firebase.app().functions('us-west2');` is called.  This is why example#6 works.

So when you then call `firebase.app().functions().httpsCallable('test')()`, the `firebase.app().functions()` call both 1) created a new instance of the functions class (since it has a different name than 'us-west2', and 2) does not set the region or emulator origin on it, causing it to call us-central1 production.

**Comment by kmcnellis** [comment]
I think it's probably working as designed, but the design seems quite confusing/unintuitive to me.

**Comment by epodol** [comment]
Thank you @kmcnellis! That is the conclusion I came to as well. I appreciate the help!

**Comment by samtstern** [comment]
@kmcnellis has nailed it. This is working as designed, and the design is confusing 🤦 

For all of our SDKs we decided that `useEmulator` would be a **per instance** setting, not a per-FirebaseApp or per-SDK setting. However each product does "instance caching" a bit differently. For Realtime Database it's per-URL caching. For Functions it's per-region.

Closing as this is not a bug, although I take it as feedback and I'll think about ways to improve this without breaking people!
